idaes/apps/grid_integration/forecaster.py

```diff
@@ -251,3 +299,412 @@ def _forecast(self, means, stds, hour, horizon, n_samples):
+
+            day_ahead_result: a Prescient RucPlan object.
+
+        Returns:
+            None
+        """
+
+        forecasts_arr = np.random.normal(
+            loc=corresponding_means, scale=corresponding_stds, size=(n_samples, horizon)
+        )
+        forecasts_arr[forecasts_arr < 0] = 0
+
+
+class Backcaster(AbstractPrescientPriceForecaster):
+
+    """
+    Generate price forecasts by directly using historical prices.
+    """
+
+    def __init__(
+        self, historical_da_prices, historical_rt_prices, max_historical_days=10
+    ):
+        """
+        Initialize the Backcaster.
+
+        Arguments:
+            historical_da_prices: dictionary of list for historical hourly day-ahead prices
+
+            historical_rt_prices: dictionary of list for historical hourly real-time prices
+
+            max_historical_days: maximum number of days of price data to store on the instance
+
+        Returns:
+            None
+        """
+
+        self.max_historical_days = max_historical_days
+        self.historical_da_prices = historical_da_prices
+        self.historical_rt_prices = historical_rt_prices
+        self._current_day_rt_prices = {bus: [] for bus in historical_da_prices}
+
+    def _validate_input_historical_price(self, historical_price):
+
+        """
+        Validate input historical prices.
+
+        Arguments:
+            historical_price: dictionary of list for historical hourly prices
+
+        Returns:
+            None
+        """
+
+        if not isinstance(historical_price, dict):
+            raise TypeError(
+                "Given historical price is not an dictionary object. Dictionaries with bus name (str) as keys are expected."
+            )
+
+        if len(historical_price) == 0:
+            raise ValueError(f"Given historical price is empty.")
+
+        for b, price_list in historical_price.items():
+            if not isinstance(price_list, list):
+                raise TypeError(
+                    f"Given historical price for bus {b} is not a list object. A list of historical prices is expected."
+                )
+
+            n_prices = len(price_list)
+
+            if n_prices < 24:
+                raise ValueError(
+                    f"At least a day of the historical prices (24 entries) is needed. For bus {b}, only {n_prices} are provided."
+                )
+
+            if n_prices % 24 != 0:
+                raise ValueError(
+                    f"The number of prices for each bus should be a multiple of 24. But for bus {b}, {n_prices} are provided."
+                )
+
+            while len(historical_price[b]) // 24 > self.max_historical_days:
+                raise Warning(
+                    f"The number of days of the provided historical prices for bus {b} is greater than the max value {self.max_historical_days}. Dropping the first day's data."
+                )
+                historical_price[b] = historical_price[b][24:]
```

Comment:
We should write a test to check this warning is raised. This would be a good "intro to testing" project for @Xinhe-Chen. @dguittet or @radhakrishnatg, are you comfortable with automated testing to show @Xinhe-Chen how to do this? (Or do it and then explain what you did.)